### PR TITLE
Extended Discovery configurability and timeouts, converge CM/AC DNS-SD fields

### LINKED ISSUE #9154: DNS-SD commissionable node discovery does not time out
#### Problem
 * default commissioning mode discovery does not time out. After 15 mins (configurable) the device should stop advertising commissionable node
 * when devices use extended discovery (eg. commissionable node advertising when not in commissioning mode), this should also time out

 #### Proposed Solution
Schedule timers from Mdns code to ensure commissionable node discovery always times out

### PR DESCRIPTION
**Pull request by chrisdecenzo**
#### Problem
What is being fixed?  Examples:
* Fix https://github.com/project-chip/connectedhomeip/issues/9154: DNS-SD commissionable node discovery does not time out
* Fix SDK to address https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/3805: DNS-SD fields - CM and AC fields are unnecessarily confusing, error prone

#### Change overview
* add timers for commissionable node discovery timeout and extended discovery timeout
* make extended discovery configurable per the spec, add shell commands for configuring
* code to match latest DNS-SD commissionable node discovery changes (CM/AC convergence)

#### Testing
How was this tested? (at least one bullet point required)
* tv-app and lighting app shell commands to toggle commissioning mode (disabled, basic, enhanced)
* chip-tool to discovery and print out DNS-SD values

### CODE REVIEW

File: src/app/server/Mdns.cpp [hunk 1]
```diff
@@ -101,6 +103,201 @@ uint16_t GetUnsecuredPort()
     return gUnsecuredPort;
 }
 
+#define TIMEOUT_CLEARED 0
+Time::TimeSource<Time::Source::kSystem> gTimeSource;
+
+int16_t gDiscoveryTimeoutSecs   = CHIP_DEVICE_CONFIG_DISCOVERY_TIMEOUT_SECS;
+uint64_t gDiscoveryExpirationMs = TIMEOUT_CLEARED;
```
Comment (msandstedt) [inline src/app/server/Mdns.cpp:110]: Can't we at least fold of these globals into the `chip::Mdns::ServiceAdvertiser::Instance()` singleton?

**Review by jelderton — APPROVED** [review]

File: src/lib/mdns/platform/tests/TestPlatform.cpp [hunk 2]
```diff
@@ -80,43 +79,73 @@ test::ExpectedCall commissionableSmall = test::ExpectedCall()
...
-                                             .AddSubtype("_T25")
-                                             .AddSubtype("_C1")
-                                             .AddSubtype("_A1");
+test::ExpectedCall commissionableLargeBasic = test::ExpectedCall()
+                                                  .SetProtocol(MdnsServiceProtocol::kMdnsProtocolUdp)
+                                                  .SetServiceName("_matterc")
+                                                  .SetHostName(host)
+                                                  .AddTxt("D", "22")
+                                                  .AddTxt("VP", "555+897")
+                                                  .AddTxt("AP", "1")
```
Comment (cecille) [inline src/lib/mdns/platform/tests/TestPlatform.cpp:104]: This should be gone, right?

File: src/lib/mdns/platform/tests/TestPlatform.cpp [hunk 3]
```diff
@@ -80,43 +79,73 @@ test::ExpectedCall commissionableSmall = test::ExpectedCall()
...
+        .SetProductId(chip::Optional<uint16_t>(897))
+        .SetRotatingId(chip::Optional<const char *>("id_that_spins"));
+
+test::ExpectedCall commissionableLargeEnhanced = test::ExpectedCall()
+                                                     .SetProtocol(MdnsServiceProtocol::kMdnsProtocolUdp)
+                                                     .SetServiceName("_matterc")
+                                                     .SetHostName(host)
+                                                     .AddTxt("D", "22")
+                                                     .AddTxt("VP", "555+897")
+                                                     .AddTxt("AP", "1")
```
Comment (cecille) [inline src/lib/mdns/platform/tests/TestPlatform.cpp:137]: ditto

File: src/app/server/Mdns.cpp (hunk 1 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: So I made chip::app::Mdns a singleton chip::app::MdnsServer.Instance() so that it can hold current app server state (which did not seem to easily belong in ServiceAdvertiser()).

File: src/app/server/Mdns.cpp [hunk 4]
```diff
@@ -101,6 +103,202 @@ uint16_t GetUnsecuredPort()
...
+    {
+        ChipLogError(Discovery, "Failed to advertise commissioner: %s", chip::ErrorStr(err));
+    }
+#endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY
+
+    gExtendedDiscoveryExpirationMs = TIMEOUT_CLEARED;
+}
+#endif // CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY
+
+void HandleDiscoveryExpiration(System::Layer * aSystemLayer, void * aAppState)
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp:164]: Added comments explaining "discovery expiration" (commissionable node advertising when in commissioning mode) and "extended discovery expiration" (commissionable node advertising when NOT in commissioning mode)

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 5]
```diff
@@ -542,7 +525,7 @@ CHIP_ERROR AdvertiserMinMdns::Advertise(const CommissionAdvertisingParameters &
 
 FullQName AdvertiserMinMdns::GetCommisioningTextEntries(const CommissionAdvertisingParameters & params)
 {
-    // Max number of TXT fields from the spec is 9: D, VP, AP, CM, DT, DN, RI, PI, PH.
+    // Max number of TXT fields from the spec is 9: D, VP, CM, DT, DN, RI, PI, PH.
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp]: good catch, fixed

File: src/lib/mdns/Advertiser_ImplMinimalMdns.cpp [hunk 6]
```diff
@@ -542,7 +525,7 @@ CHIP_ERROR AdvertiserMinMdns::Advertise(const CommissionAdvertisingParameters &
...
     constexpr size_t kMaxTxtFields = 9;
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/lib/mdns/Advertiser_ImplMinimalMdns.cpp]: good catch, fixed

File: src/lib/mdns/platform/tests/TestPlatform.cpp (hunk 2 above)
Comment (chrisdecenzo) [reply src/lib/mdns/platform/tests/TestPlatform.cpp]: good catch, fixed

File: src/lib/mdns/platform/tests/TestPlatform.cpp (hunk 3 above)
Comment (chrisdecenzo) [reply src/lib/mdns/platform/tests/TestPlatform.cpp]: good catch, fixed

**Review by bzbarsky-apple — COMMENTED** [review]
This would be a lot easier to review if the timeout bits were not lumped together with the CM/AC bits and random other changes/cleanup....

File: src/app/server/Mdns.cpp [hunk 7]
```diff
@@ -78,37 +78,204 @@ chip::ByteSpan FillMAC(uint8_t (&mac)[8])
...
+    uint64_t now = mTimeSource.GetCurrentMonotonicTimeMs();
+    if (mExtendedDiscoveryExpirationMs > now)
+    {
+        ChipLogDetail(Discovery, "HandleExtendedDiscoveryTimeout callback for reset session");
+        return;
+    }
+
+    // reset advertising
+    CHIP_ERROR err;
+    err = chip::Mdns::ServiceAdvertiser::Instance().StopPublishDevice();
```
Comment (msandstedt) [inline src/app/server/Mdns.cpp:132]: Is there a reason we are forced to stop advertisement of all services just to remove one service (in this case, the commissionable node advertisement)?  I do see ChipMdnsPublishService that has no symmetric ChipMdnsUnpublishService method.  Is there something architecturally that forces us into this situation where we can write individual services, but have to delete them all just to remove one?

This seems inherently fragile.  For instance, what if the operational advertisement needs to change during commissionable node advertisement?  Won't that break this state machine that's operating through these timer callbacks?

It seems like this code would be far simpler if we could selectively publish and unpublish each service.

**Review by msandstedt — COMMENTED** [review]
I had a comment inline.  I'm wondering what the fundamental limitation is that prevents us from unpublishing individual services.  Apparently, we can only delete them all at once.

That is introducing a lot of complication (and likely fragility) here.

Other than that, the PR appears to do what it purports to do and makes sense.

File: src/app/server/Mdns.cpp (hunk 7 above)
Comment (cecille) [reply src/app/server/Mdns.cpp:132]: We did have a plan to have Stop commands available for ex operational and commissionable and commissioner separately. So this is the plan, we're just not there yet. There are a couple of things that need to happen before that happens.
1) The platform mdns API needs to track all the services it actually started. It's coming, but it relies on a couple of other PRs that are also coming.
2) The platform impls actually need implementations for their individual stop commands. Right now they're all marked as not implemented.

The TL;DR is you're right and it's coming, but it's more work than it sounds like and this PR is working with what it has.

Comment (msandstedt) [reply src/app/server/Mdns.cpp:132]: Got it.  Thank you.

**Review by msandstedt — APPROVED** [review]

**Review by cecille — APPROVED** [review]
Does anyone recognize the cirque error?

otbr-agent[55]: [CRIT]-PLAT----: Init() at ../../third_party/openthread/repo/src/posix/platform/hdlc_interface.cpp:150: InvalidArgument

I've seen this on another PR recently, but I'm not familiar with that area of the code. 

As far as I can tell, the cirque failure is unrelated to this PR. CRA/CRI change looks good. I'm less familar with the spec about the timing, but it looks correct from my reading.

**Comment by chrisdecenzo** [comment]
> Does anyone recognize the cirque error?
> 
> otbr-agent[55]: [CRIT]-PLAT----: Init() at ../../third_party/openthread/repo/src/posix/platform/hdlc_interface.cpp:150: InvalidArgument
> 
> I've seen this on another PR recently, but I'm not familiar with that area of the code.
> 
> As far as I can tell, the cirque failure is unrelated to this PR. CRA/CRI change looks good. I'm less familar with the spec about the timing, but it looks correct from my reading.

I couldn't figure out why this would have anything to do with this PR. I'll check with the larger group on whether this is known

**Review by sharadb-amazon — APPROVED** [review]

File: examples/minimal-mdns/advertiser.cpp [hunk 8]
```diff
@@ -124,10 +123,15 @@ bool HandleOptions(const char * aProgram, OptionSet * aOptions, int aIdentifier,
         gOptions.productId = Optional<uint16_t>::Value(static_cast<uint16_t>(atoi(aValue)));
         return true;
     case kOptionCommisioningMode:
-        gOptions.commissioningMode = true;
-        return true;
-    case kOptionCommisioningModeOpenWindow:
-        gOptions.commissioningModeOpenWindow = true;
+        cm = static_cast<uint8_t>(atoi(aValue));
```
Comment (andy31415) [inline examples/minimal-mdns/advertiser.cpp:126]: nit: prefer strcmp with values. I would rather read `--commissioning-mode basic` or `--commissioning-mode enhanced` than `--commissioning-mode 1` or `--commissioning-mode 2`. A log saying 'invalid value' is better than return without change (I would not know that `--commissioning-mode 123` does not work)

I believe eventually the minimal-dns examples may go away though, so will merge even if this is not done in this PR.

File: examples/minimal-mdns/client.cpp [hunk 9]
```diff
@@ -309,14 +309,15 @@ int main(int argc, char ** args)
 
     mdns::Minimal::Server<20> mdnsServer;
     ReportDelegate reporter;
+    CHIP_ERROR err;
```
Comment (andy31415) [inline examples/minimal-mdns/client.cpp:312]: why this change? This increases the scope of err

File: src/app/server/Mdns.cpp [hunk 10]
```diff
@@ -78,37 +78,204 @@ chip::ByteSpan FillMAC(uint8_t (&mac)[8])
...
+    {
+        ChipLogError(Discovery, "Failed to get extended timeout configuration err: %s", chip::ErrorStr(err));
+        secs = CHIP_DEVICE_CONFIG_EXTENDED_DISCOVERY_TIMEOUT_SECS;
+    }
+
+    ChipLogDetail(Discovery, "GetExtendedDiscoveryTimeoutSecs %d", secs);
+    return secs;
+}
+
+/// Callback from Extended Discovery Expiration timer
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:107]: doccomments should be in the header not cpp file (when people see the interface they use, they read the header)

File: src/app/server/Mdns.cpp [hunk 11]
```diff
@@ -78,37 +78,204 @@ chip::ByteSpan FillMAC(uint8_t (&mac)[8])
...
+{
+    MdnsServer::Instance().OnExtendedDiscoveryExpiration(aSystemLayer, aAppState);
+}
+
+/// Checks if extended discovery has expired and if so,
+/// stops commissionable node advertising
+/// Extended Discovery Expiration refers here to commissionable node advertising when NOT in commissioning mode
+void MdnsServer::OnExtendedDiscoveryExpiration(System::Layer * aSystemLayer, void * aAppState)
+{
+    if (mExtendedDiscoveryExpirationMs == TIMEOUT_CLEARED)
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:118]: This seems to share a lot of code in common with the 'OnDiscoveryExpiration' below.
Could you share the code?

File: src/app/server/Mdns.cpp [hunk 12]
```diff
@@ -78,37 +78,204 @@ chip::ByteSpan FillMAC(uint8_t (&mac)[8])
...
-    gUnsecuredPort = port;
+    if (mDiscoveryTimeoutSecs == CHIP_DEVICE_CONFIG_DISCOVERY_NO_TIMEOUT)
+    {
+        return CHIP_NO_ERROR;
+    }
+    ChipLogDetail(Discovery, "Scheduling Discovery timeout in secs=%d", mDiscoveryTimeoutSecs);
+
+    mDiscoveryExpirationMs = mTimeSource.GetCurrentMonotonicTimeMs() + static_cast<uint64_t>(mDiscoveryTimeoutSecs) * 1000;
+
+    ReturnErrorOnFailure(DeviceLayer::SystemLayer().StartTimer(static_cast<uint32_t>(mDiscoveryTimeoutSecs) * 1000,
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:245]: nit: you could directly

```cpp
return DeviceLayer::SystemLayer().StartTimer(static_cast<uint32_t>(mDiscoveryTimeoutSecs) * 1000,
                                             HandleDiscoveryExpiration, nullptr));
```

**Review by andy31415 — APPROVED** [review]

File: examples/minimal-mdns/client.cpp (hunk 9 above)
Comment (chrisdecenzo) [reply examples/minimal-mdns/client.cpp:312]: just trying to fix the build

File: src/app/server/Mdns.cpp (hunk 10 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp:107]: this method is not in the header, but for the others I moved the comments

File: src/app/server/Mdns.cpp (hunk 11 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: created a shared method to call

File: src/app/server/Mdns.cpp (hunk 12 above)
Comment (chrisdecenzo) [reply src/app/server/Mdns.cpp]: fixed in 2 places

File: examples/minimal-mdns/advertiser.cpp (hunk 8 above)
Comment (chrisdecenzo) [reply examples/minimal-mdns/advertiser.cpp:126]: leaving this for now

**Comment by mspang** [comment]
This PR broke commissioning ESP32 using chip-tool. ESP32 gets on the network and can be discovered but chip-tool never actually discovers it, and instead hangs until the whole process times out.

Repro:
chip-tool pairing ble-wifi <ssid> <psk> 112233 20202021 3840

File: src/app/server/Mdns.h [hunk 13]
```diff
@@ -18,50 +18,119 @@
...
+#include <platform/CHIPDeviceLayer.h>
 #include <stddef.h>
+#include <system/TimeSource.h>
 
 namespace chip {
 namespace app {
-namespace Mdns {
 
-enum class CommissioningMode
+#define TIMEOUT_CLEARED 0
```
Comment (mspang) [inline src/app/server/Mdns.h:29]: Can we remove this namespace pollution please?

File: src/app/server/Mdns.cpp [hunk 14]
```diff
@@ -255,66 +381,82 @@ CHIP_ERROR Advertise(bool commissionableNode, CommissioningMode mode)
...
+void MdnsServer::StartServer(chip::Mdns::CommissioningMode mode)
 {
-    CHIP_ERROR err = chip::Mdns::ServiceAdvertiser::Instance().Start(&chip::DeviceLayer::InetLayer, chip::Mdns::kMdnsPort);
+    ChipLogDetail(Discovery, "Mdns StartServer mode=%d", static_cast<int>(mode));
 
-    err = app::Mdns::AdvertiseOperational();
+    ClearTimeouts();
+
+    CHIP_ERROR err;
+    err = chip::Mdns::ServiceAdvertiser::Instance().StopPublishDevice();
```
Comment (mspang) [inline src/app/server/Mdns.cpp:401]: This is the change that broke commissioning.

This removes all of the records, then restarts the server. Restarting the server is the only time unsolicited advertisements are sent. It is generally always the case that the device gets onto the network after the initial query and thus it is the unsolicited advertisements that provided the answer (via AdvertiseRecords()).
